File: paiza/python/DP/Leetcode/322.coin_change.py
from typing import List

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        
        # initialize
        dp = [[float('inf')] * (amount+1) for _ in range(len(coins))]
        for i in range(amount+1):
            if i % coins[0] == 0:
                dp[0][i] = int(i / coins[0])

        for i in range(1, len(coins)):
            for j in range(amount+1):

                if dp[i-1][j] != float('inf'):
                    dp[i][j] = min(dp[i][j], dp[i-1][j])

                if j - coins[i] < 0:
                    continue

                if dp[i][j-coins[i]] != float('inf'):
                    dp[i][j] = min(dp[i][j], dp[i][j-coins[i]] + 1)

        if dp[len(coins)-1][amount] == float('inf'):
            return -1
        else:
            return dp[len(coins)-1][amount]

sol = Solution()
print(sol.coinChange([1,2,5], 11))
print(sol.coinChange([1],0))

File: Leetcode_bests100/DP/322.coin_change.java
class Main{
  public static void main(String[] args){
    Solution sl = new Solution();
    System.out.println(sl.coinChange(new int[] {1,2,5}, 11));
    System.out.println(sl.coinChange(new int[] {2}, 3));
    System.out.println(sl.coinChange(new int[] {1}, 0));


  }
}

class Solution{
  public int coinChange(int[] coins, int amount){
    int[] dp = new int[amount+1];
    for (int i=0; i<dp.length; i++){
      dp[i] = Integer.MAX_VALUE-1;
    }
    dp[0] = 0;

    for (int i=1; i<amount+1; i++){
      for (int j=0; j<coins.length; j++){
        if (coins[j] > i){
          continue;
        }
        dp[i] = Math.min(dp[i], dp[i-coins[j]]+1);
      }
    }

    if (dp[dp.length-1] == Integer.MAX_VALUE-1){
      return -1;
    }

    return dp[dp.length-1];
  }
}